### PR TITLE
bpo-36012: Avoid linear slot search when it's easy to know that we are not looking for a slot name

### PR DESCRIPTION
**Pull request by scoder**
Speeds up class attribute setting for non-slot names by 50%.

https://bugs.python.org/issue36012

### CODE REVIEW

File: Objects/typeobject.c [hunk 1]
```diff
@@ -7240,6 +7240,11 @@ update_slot(PyTypeObject *type, PyObject *name)
        recursing into subclasses. */
     PyType_Modified(type);
 
+    if (PyUnicode_READ_CHAR(name, 0) != '_') {
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: Could you check this before calling `update_slot()`? In `update_all_slots()` it is called with all dunder names.

It may be worth to perform more strict check and speed up also setting protected and private attributes.

Comment (scoder) [reply Objects/typeobject.c]: Probably yes, preferably in a little helper function in that case, like `looks_like_special_name()`.
But wouldn't we still have to call `PyType_Modified()`?

Comment (scoder) [reply Objects/typeobject.c]: I implemented a tighter check. Timings are pretty much the same.

File: Objects/typeobject.c [hunk 2]
```diff
@@ -7000,6 +7005,34 @@ static slotdef slotdefs[] = {
...
 };
 
+
+/* Since searching slotdefs[] is expensive and assigning to slots is rare,
+   do a quick probability check if the PyUnicode name could be a slot name
+   at all. */
+static int name_looks_like_slot(PyObject *name) {
+    Py_ssize_t length = PyUnicode_GET_LENGTH(name);
+    /* The longest slot name is currently "__getattribute__": 16 characters */
+    if (length >= 6 && length <= 16) {
```
Comment (pablogsal) [inline Objects/typeobject.c]: I don't like this. If in the future a new slot is added we need to remember to update this line and I am almost certain it will be missed. 

IMO, either this is calculated at compilation time based on `slotsdef` or should be removed.

File: Objects/typeobject.c [hunk 3]
```diff
@@ -7000,6 +7005,34 @@ static slotdef slotdefs[] = {
...
+        /* All slot names are pure ASCII. */
+        int kind = PyUnicode_KIND(name);
+        if (kind == PyUnicode_1BYTE_KIND) {
+            void *data = PyUnicode_DATA(name);
+            /* Allow the C compiler to read 2 bytes instead of two times 1 byte. */
+            if ((PyUnicode_READ(kind, data, 0) == '_') &
```
Comment (pablogsal) [inline Objects/typeobject.c]: Also, to make this a bit more clear, could we invert every check and do early returns?

I think that makes the function more clear and reduces a lot the indentation levels.

File: Objects/typeobject.c [hunk 4]
```diff
@@ -2108,6 +2108,7 @@ static int update_slot(PyTypeObject *, PyObject *);
 static void fixup_slot_dispatchers(PyTypeObject *);
 static int set_names(PyTypeObject *);
 static int init_subclass(PyTypeObject *, PyObject *);
+static int name_looks_like_slot(PyObject *name);
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: If define it before use you will not need a forward declaration.

I would name it `is_dunder_name`.

File: Objects/typeobject.c [hunk 5]
```diff
@@ -3289,7 +3290,11 @@ type_setattro(PyTypeObject *type, PyObject *name, PyObject *value)
     }
     res = _PyObject_GenericSetAttrWithDict((PyObject *)type, name, value, NULL);
     if (res == 0) {
-        res = update_slot(type, name);
+        if (name_looks_like_slot(name)) {
+            res = update_slot(type, name);
+        } else {
+            PyType_Modified(type);
```
Comment (serhiy-storchaka) [inline Objects/typeobject.c]: Maybe move `PyType_Modified()` out of `update_slot()`? Then it could be called unconditionally here and only once in `update_all_slots()`.

File: Objects/typeobject.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c]: I would just merge two ifs.

Parenthesis are not needed around the `==` operator. IMHO they reduce readability.

Comment (scoder) [reply Objects/typeobject.c]: For me, they increase readability, because they make the separate conditions explicit. Also, C compilers often warn about "&" in plain boolean expressions, because its usage can hint at a typo.
I'll reduce the nesting, though.

File: Objects/typeobject.c (hunk 4 above)
Comment (scoder) [reply Objects/typeobject.c]: I've implemented it in that place because that keeps it close to the list of slot names list, so that it's less likely that the two get out of sync somehow.

File: Objects/typeobject.c (hunk 5 above)
Comment (scoder) [reply Objects/typeobject.c]: I think that's a good idea.

File: Objects/typeobject.c (hunk 2 above)
Comment (scoder) [reply Objects/typeobject.c]: I was thinking that, too, that's why I implemented the function here and added the comment. But I can live with removing the maximum length check.

**Comment by scoder** [comment]
I think I resolved all review comments.

File: Objects/typeobject.c (hunk 3 above)
Comment (serhiy-storchaka) [reply Objects/typeobject.c]: Is not there a typo? Perhaps you meant "&&"?

Comment (scoder) [reply Objects/typeobject.c]: Read the comment above the line. :)
When I write `a[0] == 1 && a[1] == 2 && a[2] == 3 && ...`, then the C compiler has to assume that there is a data dependency between the three items (`a[1]` might only be readable if `a[0] == 1`, and might otherwise segfault – think of 0-terminated C strings), so it must generate slow and branchy code that reads one byte at a time.
When, instead, I write `(a[0] == 1) & (a[1] == 2) & (a[2] == 3) & (a[3] == 4) & ...`, then the C compiler can safely assume that all fields are accessible and that there will be no segfault, so that it can generate a single memory access and comparison for it, e.g. reading ints instead of bytes.
Basically, I'm leaving it to the C compiler to do the right thing, instead of preventing it from taking the best decision.

Comment (pablogsal) [reply Objects/typeobject.c]: >Read the comment above the line. :)
When I write `a[0] == 1 && a[1] == 2 && a[2] == 3 && ...`, then the C compiler has to assume that there is a data dependency between the three items (`a[1]` might only be readable if `a[0] == 1`, and might otherwise segfault – think of 0-terminated C strings), so it must generate slow and branchy code that reads one byte at a time.
When, instead, I write `(a[0] == 1) & (a[1] == 2) & (a[2] == 3) & (a[3] == 4) & ...`, then the C compiler can safely assume that all fields are accessible and that there will be no segfault, so that it can generate a single memory access and comparison for it, e.g. reading ints instead of bytes.
Basically, I'm leaving it to the C compiler to do the right thing, instead of preventing it from taking the best decision.

I am not very sure about this assertion. This is the x86-64 generated for the two options:

https://godbolt.org/z/SgXbL9

Unless I am missing something, both generate the same branches and I would even say that the 2nd version generates more efficient asm (this probably will be very susceptible to other aspects of the surrounding code so I don't take this very seriously :) ). I would also want to run some benchmarks to check if there is any difference worth the surprise when reading & instead of &&.

I know it sounds cliché but in general, is best not the help the compiler. This is specially true when running with very different set of compilers/platforms.

Comment (scoder) [reply Objects/typeobject.c]: I played around with the code in godbolt and I think I found a reason for this. gcc seems to consider a word access more costly than a byte access. You can see that when you cast the byte pointer to a short pointer to compare the first 16 bits in one go. It then moves the start word access after the end byte access when allowed by the "&".

Anyway, I see the argument of least surprise. I'll use "&&". Thanks for insisting.

File: Objects/typeobject.c [hunk 6]
```diff
@@ -3164,6 +3164,25 @@ _PyType_LookupId(PyTypeObject *type, struct _Py_Identifier *name)
...
+{
+    Py_ssize_t length = PyUnicode_GET_LENGTH(name);
+    int kind = PyUnicode_KIND(name);
+    /* Special names contain at least "__x__" and are always ASCII. */
+    if (length > 4 && kind == PyUnicode_1BYTE_KIND) {
+        Py_UCS1 *characters = PyUnicode_1BYTE_DATA(name);
+        /* Allow the C compiler to read 2 bytes instead of two times 1 byte. */
+        return (
+            ((characters[0] == '_') & (characters[1] == '_')) &
+            ((characters[length-2] == '_') & (characters[length-1] == '_'))
```
Comment (rhettinger) [inline Objects/typeobject.c]: This all looks fine.  One nit, I would take a hint from string search algorithms and reorder the tests to those most likely to fail first.  Instead of `(0, 1, -2 -1)`, I would use `(-2, -1,  0, 1)`.

File: Objects/typeobject.c [hunk 7]
```diff
@@ -7000,6 +7028,7 @@ static slotdef slotdefs[] = {
     {NULL}
 };
 
+
```
Comment (rhettinger) [inline Objects/typeobject.c]: Remove spurious newline.

File: Objects/typeobject.c [hunk 8]
```diff
@@ -7278,6 +7300,9 @@ update_all_slots(PyTypeObject* type)
 {
     slotdef *p;
 
+    /* Clear the VALID_VERSION flag of 'type' and all its subclasses. */
+    PyType_Modified(type);
```
Comment (rhettinger) [inline Objects/typeobject.c:7304]: Was this intended?  I don't see how it is related to the dunder method test.

File: Objects/typeobject.c [hunk 9]
```diff
@@ -3164,6 +3164,25 @@ _PyType_LookupId(PyTypeObject *type, struct _Py_Identifier *name)
...
+   is a double-underscore special name. */
+static int
+is_dunder_name(PyObject *name)
+{
+    Py_ssize_t length = PyUnicode_GET_LENGTH(name);
+    int kind = PyUnicode_KIND(name);
+    /* Special names contain at least "__x__" and are always ASCII. */
+    if (length > 4 && kind == PyUnicode_1BYTE_KIND) {
+        Py_UCS1 *characters = PyUnicode_1BYTE_DATA(name);
+        /* Allow the C compiler to read 2 bytes instead of two times 1 byte. */
```
Comment (rhettinger) [inline Objects/typeobject.c]: I'm not sure why this comment is here.   What would cause it to reader two bytes at a time?

There GCC-8 disassembly shows fetch and compares one byte at a time:

```
L4922:
    cmpb    $95, (%rax)
    jne L4918
    cmpb    $95, 1(%rax)
    je  L4940
    ...

L4940:
    cmpb    $95, -2(%rax,%rcx)
    jne L4918
    cmpb    $95, -1(%rax,%rcx)
    jne L4918
    movq    %rdi, %rsi
    movq    %rbx, %rdi
    call    _update_slot
    movl    %eax, %ebp
```

**Review by rhettinger — CHANGES_REQUESTED** [review]

File: Objects/typeobject.c (hunk 8 above)
Comment (scoder) [reply Objects/typeobject.c:7304]: Yes, Serhiy and me agreed that it would be a good idea to move this out of `update_slot()` because it needs to be called unconditionally, and currently it's also called repeatedly for each slot in `update_all_slots()`. Now, both callers of `update_slot()` call `PyType_Modified()` directly, unconditionally and only once.

File: Objects/typeobject.c (hunk 9 above)
Comment (scoder) [reply Objects/typeobject.c]: This has already been changed.

File: Objects/typeobject.c (hunk 7 above)
Comment (scoder) [reply Objects/typeobject.c]: Done

File: Objects/typeobject.c (hunk 6 above)
Comment (scoder) [reply Objects/typeobject.c]: Good idea. Done.

File: Objects/typeobject.c (hunk 8 above)
Comment (rhettinger) [reply Objects/typeobject.c:7304]: Thanks.

File: Doc/whatsnew/3.8.rst [hunk 10]
```diff
@@ -375,6 +375,11 @@ Optimizations
   This makes the created list 12% smaller on average. (Contributed by
   Raymond Hettinger and Pablo Galindo in :issue:`33234`.)
 
+* Doubled the speed of class variable writes.  When a non-dunder attribute
+  was updated, there was an unnecessary call to update slots.
+  (Contributed by Stefan Behnel, Pablo Galindo Salgado, Raymond Hettinger,
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.8.rst:380]: No need to mentioned me here. I just left few review comments. The author of the idea and the implementation is Stefan.

**Review by rhettinger — APPROVED** [review]

File: Doc/whatsnew/3.8.rst (hunk 10 above)
Comment (scoder) [reply Doc/whatsnew/3.8.rst:380]: I'm happy with having all the names in there. It was a joint effort, with lots of good reviews.
